Tests/test_map.gd

```diff
@@ -12,3 +12,125 @@ func test_map_gen() -> void:
+
+
+## Rule 3: There must be at least 2 room types among destinations of Rooms that have 2 or more Paths going out.
+func test_rule_3_map_gen() -> void:
+	var queue : Array[String] = []
+	var unique: Array[String] = []
+	var current_room: RoomBase
+	# Position of the rightmost room of the floor
+	var rightmost_room_position: Vector2
+	
+	# Last floor is Boss room, second last floor is heal rooms
+	for index_height: int in range(test_map.rooms.size()-2):
+		# Reset queue when new floor
+		queue = []
+		
+		for index_width: int in range(test_map.rooms[index_height].size()):
+			current_room = test_map.rooms[index_height][index_width]
+			
+			if current_room == null:
+				continue
+			
+			# If current_room is not null, rightmost_room_position is updated
+			rightmost_room_position = current_room.room_position
+			queue.append(current_room.room_event.get_event_name())
+			
+			while queue.size() > 3:
+				queue.pop_front()
+				
+			if queue.size() >= 2:
+				unique = []
+
+				# Remove duplicates
+				for event in queue:
+					if not unique.has(event):
+						unique.append(event)
+				
+				if queue.size() == 2:
+					assert_true(unique.size() != 1, "The two leftmost rooms of a floor should not have the same event but found %s at %s and %s at %s " % 
+					[test_map.rooms[index_height][index_width-1].room_event.get_event_name(), test_map.rooms[index_height][index_width-1].room_position, 
+					 test_map.rooms[index_height][index_width].room_event.get_event_name(), test_map.rooms[index_height][index_width].room_position])
+
+				elif queue.size() == 3:
+					assert_true(unique.size() != 1, "3 side by side rooms should not have the same event but found %s at %s, %s at %s and %s at %s " % 
+					[test_map.rooms[index_height][index_width-2].room_event.get_event_name(), test_map.rooms[index_height][index_width-2].room_position,
+					 test_map.rooms[index_height][index_width-1].room_event.get_event_name(), test_map.rooms[index_height][index_width-1].room_position, 
+					 test_map.rooms[index_height][index_width].room_event.get_event_name(), test_map.rooms[index_height][index_width].room_position])
+
+		# Check if the two rightmost rooms of the floor are the same type or not
+		queue.pop_front()
+		if queue.size() >= 2:
+			unique = []
+
+			# Remove duplicates
+			for event in queue:
+				if not unique.has(event):
+					unique.append(event)
+			
+			assert_true(unique.size() != 1, "The two rightmost rooms of the floor should not have the same event but found %s at %s and %s at %s " % 
+					[test_map.rooms[rightmost_room_position.y][rightmost_room_position.x-1].room_event.get_event_name(), test_map.rooms[rightmost_room_position.y][rightmost_room_position.x-1].room_position, 
+					 test_map.rooms[rightmost_room_position.y][rightmost_room_position.x].room_event.get_event_name(), test_map.rooms[rightmost_room_position.y][rightmost_room_position.x].room_position])
```

Comment:
isn't that already covered by the main checking loop ? when you get to the right, you end up on null rooms, so you stop appending but you pop the one at x-2, so you just check the two rightmost rooms at the end of the loop, no ?

Reply:
I've put a `continue` so it just skips when it reaches a null room and doesn't pop the queue, so the queue stays at size 3 until the end of the `for` loop. I'll check to be sure but I remember that's why I put that part

Reply:
Putting the pop before the continue might solve this problem